src/Npgsql/Internal/TypeHandlers/JsonHandler.cs

```diff
@@ -71,6 +75,20 @@ protected internal override int ValidateAndGetLengthCustom<TAny>([DisallowNull]
                 parameter.ConvertedValue = data;
             return lengthCache.Set(data.Length + _headerLen);
         }
+        
+#if NET6_0_OR_GREATER || NETSTANDARD2_0 || NETSTANDARD2_1
+        if (typeof(TAny) == typeof(JsonObject) || typeof(TAny) == typeof(JsonArray))
+        {
+            lengthCache ??= new NpgsqlLengthCache(1);
+            if (lengthCache.IsPopulated)
+                return lengthCache.Get();
+
+            var data = SerializeJsonObject((JsonNode)(object)value!);
```

Comment:
nit: I think the bang isn't needed (here and for the other types) - the null case is supposed to be handled outside of WriteWithLengthCustom